### PR TITLE
Issue "Cannot find name did-you-mean" errors as suggestions in plain JS

### LINKED ISSUE #41582: Provide high-confidence spelling suggestions as suggestion diagnostics
When JavaScript users don't have type-checking turned on, there are often some good suggestions left on the table that go to waste.

Obviously we can't just provide all the errors that `checkJs` would issue, but there are a few good ones. Specifically, spelling suggestions could be provided if we were more confident in how useful they are, where confidence is determined by some rules.

Some of the following rules might be worth investigating:

* the suggestion occurs in a module file
* the "misspelled" identifier occurs nowhere else in the program
* the "misspelled" identifier is not a commonly-used identifier in well-known libraries
* the suggestion's edit distance is lower than the usual threshold

### PR DESCRIPTION
**Pull request by sandersn**
This PR issues "cannot find ${name}, did you mean ${name}" errors for identifiers and propery access expressions in JS files *without* `// @ts-check` and without `// @ts-nocheck`. This brings some benefits of Typescript's binder to all Javascript users, even those who haven't opted into Typescript checking.

In this PR, the errors are provided as suggestions. If the suggestions are useful, future work will change the presentation to be more obvious.

```js
export var inModule = 1
inmodule.toFixed() // errors on exports

function f() {
    var locals = 2
    locale.toFixed() // errors on locals
}
var object = {
    spaaace: 3
}
object.spaaaace // error on read
object.spaace = 2 // error on write
object.fresh = 12 // OK, no spelling correction to offer
```

To disable the errors, add `// @ts-nocheck` to the file. To get the normal checkJs experience, add `// @ts-check`.

## Why This Works ##

In a word: precision. This change has low recall &mdash; it misses lots of correct errors that would be nice to show &mdash; but it has high
precision: almost all the errors it shows are correct. And they come with a suggested correction.

Here are the ingredients:

1. For unchecked JS files, the compiler suppresses all errors except two did-you-mean name resolution errors.
2. Did-you-mean spelling correction is already tuned for high precision/low recall, and doesn't show many bogus errors even in JS.
3. The error is suppressed for suggestions from global files that are not the current file. These are often DOM feature detection, for example.
4. The error is suppressed for property accesses on a class because the compiler doesn't understand JS constructor functions well enough. In particular, it doesn't understand any inheritance patterns.
5. The error is suppressed for `this` property accesses for the same reason, and because some `this` property accesses in JS are not on classes.

## Results ##

I tested on a random sampling of 1000 single files in a fully installed clone of Definitely Typed. I didn't get false positives from these new errors.
I also tested on all the typescript user tests that are written in TS. There, the results informed the above rules. I got the following results

### Good ###

- minimatch: Found a 9-year-old bug in untested code that has never worked.

### Bad ###

- grunt: Didn't understand grunt's convenience `require` wrapper.
- debug: Didn't produce an index signature for a loop in a `setup` function that copies properties from one object to another.
- Uglify-JS: Didn't understand post-initialisation addition of properties to a very large object &mdash; because the object was large, the compiler suggested a similar name already in it.
- chrome-devtools-frontend: Didn't understand lazy initialization of properties in several classes. The access was  '_property', the suggestion was 'property'.

## Open Questions ##

- I'm not sure of the best way to indicate that getSemanticDiagnostics should return errors for unchecked JS files. I threaded an optional boolean parameter through the interface for now.
- The wording for the error might need to be customised to point new users to the codefix.
- VS, for example, doesn't use tsconfig as much and may need a user preference to disable these errors, since it won't be easy to add `checkJs: false` to a tsconfig that doesn't exist.

## Future Work ##

I think there are a lot of other errors that we can issue in JS with enough precision. My next step is to find them and test them.

Fixes #41582

### CODE REVIEW

**Comment by sandersn** [comment]
@typescript-bot pack this

File: src/compiler/builder.ts [hunk 1]
```diff
@@ -693,7 +693,7 @@ namespace ts {
             const cachedDiagnostics = state.semanticDiagnosticsPerFile.get(path);
             // Report the bind and check diagnostics from the cache if we already have those diagnostics present
             if (cachedDiagnostics) {
-                return filterSemanticDiagnotics(cachedDiagnostics, state.compilerOptions);
```
Comment (sandersn) [inline src/compiler/builder.ts:696]: driveby typo fix

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -2059,10 +2059,13 @@ namespace ts {
                         let suggestion: Symbol | undefined;
                         if (suggestedNameNotFoundMessage && suggestionCount < maximumSuggestionCount) {
                             suggestion = getSuggestedSymbolForNonexistentSymbol(originalLocation, name, meaning);
-                            const isGlobalScopeAugmentationDeclaration = suggestion && suggestion.valueDeclaration && isAmbientModule(suggestion.valueDeclaration) && isGlobalScopeAugmentation(suggestion.valueDeclaration);
+                            const isGlobalScopeAugmentationDeclaration = suggestion?.valueDeclaration && isAmbientModule(suggestion.valueDeclaration) && isGlobalScopeAugmentation(suggestion.valueDeclaration);
```
Comment (sandersn) [inline src/compiler/checker.ts:2062]: driveby upgrade to `?.`

File: src/compiler/program.ts [hunk 3]
```diff
@@ -1915,15 +1916,27 @@ namespace ts {
 
                 Debug.assert(!!sourceFile.bindDiagnostics);
 
-                const isCheckJs = isCheckJsEnabledForFile(sourceFile, options);
+                const isCheckJsTrue = isCheckJsEnabledForFile(sourceFile, options);
+                const isCheckJsUndefined = includeUncheckedJS
+                    && sourceFile.checkJsDirective === undefined && options.checkJs === undefined
+                    && (sourceFile.scriptKind === ScriptKind.JS || sourceFile.scriptKind === ScriptKind.JSX);
```
Comment (sandersn) [inline src/compiler/program.ts:1922]: github does a poor job diffing in this function. `isCheckJsUndefined` is the only real addition here, in `includeBindAndCheckDiagnostics`, and the new `if (isCheckJsUndefined) ...`

I also renamed `isCheckJs` -> `isCheckJsTrue`, which is what confuses the differ.

File: src/compiler/program.ts [hunk 4]
```diff
@@ -1832,7 +1833,7 @@ namespace ts {
         }
 
         function getBindAndCheckDiagnostics(sourceFile: SourceFile, cancellationToken?: CancellationToken): readonly Diagnostic[] {
-            return getBindAndCheckDiagnosticsForFile(sourceFile, cancellationToken);
+            return getBindAndCheckDiagnosticsForFile(sourceFile, cancellationToken, /*includeUncheckedJS*/ undefined);
```
Comment (sheetalkamat) [inline src/compiler/program.ts:1836]: Having two different behavior between LS and command line could also mean that we cant use tsbuildinfo cached semantic diagnostics if and when we start using tsbuildinfo to report errors on closed files.

Comment (sandersn) [reply src/compiler/program.ts:1836]: I can think of a couple ways to fix this
- Always return all errors, then have `tsc` have a filter when displaying errors for JS files. This would add errors to all other API consumers, though.
- Give the language service a separate way to request unchecked JS errors.  The checker will have to store these errors separately and the language service will have to add them late.
- Possibly: Not report these errors in the "errors in closed files" list. But I don't know if that would be what users of that feature expect.

Opinions? Ideas?

I'm pretty sure that this feature is going to stick around and grow, so I'd like for it to be easier to use than to not use. Specifically, it's annoying if the future status quo for the language service is to request diagnostics twice, once for unchecked JS and again for everything else.

Comment (sheetalkamat) [reply src/compiler/program.ts:1836]: > Give the language service a separate way to request unchecked JS errors. The checker will have to store these errors separately and the language service will have to add them late.

Isn't that what suggestion diagnostics are suppose to be and seems like a good place to add those errors ?

Comment (sandersn) [reply src/compiler/program.ts:1836]: They are quite similar, and that would work except that the display should be squiggly underlines, like an error, instead of three grey dots, like a suggestion.

**Comment by DanielRosenwasser** [comment]
@typescript-bot pack this

File: src/compiler/program.ts [hunk 5]
```diff
@@ -1804,25 +1804,26 @@ namespace ts {
 
         function getDiagnosticsHelper<T extends Diagnostic>(
             sourceFile: SourceFile | undefined,
-            getDiagnostics: (sourceFile: SourceFile, cancellationToken: CancellationToken | undefined) => readonly T[],
-            cancellationToken: CancellationToken | undefined): readonly T[] {
+            getDiagnostics: (sourceFile: SourceFile, cancellationToken: CancellationToken | undefined, includeUncheckedJS: boolean | undefined) => readonly T[],
+            cancellationToken: CancellationToken | undefined,
+            includeUncheckedJS: boolean | undefined): readonly T[] {
```
Comment (uniqueiniquity) [inline src/compiler/program.ts:1809]: Why is it preferable to add this to getDiagnosticsHelper instead of just passing it through to getSemanticDiagnosticsForFile in getSemanticDiagnostics?

File: src/compiler/program.ts (hunk 4 above)
Comment (sandersn) [reply src/compiler/program.ts:1836]: @mjbvz Would VS Code be OK to add a separate call to a new function `getUncheckedJSDiagnostics` with a matching protocol message `uncheckedJSDiagnosticsSync`? It would make it easier to create a setting to disable the errors and for Code to style them differently if desired.

Also @uniqueiniquity @jessetrinity you probably have opinions from the VS side.

**Comment by sandersn** [comment]
@DanielRosenwasser you'll have to test this locally; I found that the playground always has `checkJs: true` for JS files, 

(It's a great default since it shows off our JS checking, but makes it impossible to test this PR.)

File: src/compiler/program.ts (hunk 5 above)
Comment (sandersn) [reply src/compiler/program.ts]: No good reason, I was just following compile errors around and mechanically adding parameters.

File: src/compiler/program.ts (hunk 4 above)
Comment (uniqueiniquity) [reply src/compiler/program.ts:1835]: We certainly can add an additional call on the VS side.
I'm not sure I understand why it's inappropriate for `tsc` to report these, though?

Comment (sheetalkamat) [reply src/compiler/program.ts:1835]: @uniqueiniquity is right, why should the behaviour be different between LS and tsc ?

Comment (DanielRosenwasser) [reply src/compiler/program.ts:1835]: I do think treading lightly might be the way to go - we've never put errors in JS files unless they're specifically parsing errors.

Providing these as suggestion diagnostics keeps these as "quiet" by default. We can then potentially have nightly/insiders editors promote these to green or red squiggles (similar to how they specially-handle unused variable diagnostics) and see if we get any feedback on it. What do you all think? (CC @mjbvz @RyanCavanaugh)

Comment (sandersn) [reply src/compiler/program.ts:1835]: @uniqueiniquity @sheetalkamat 
> I'm not sure I understand why it's inappropriate for tsc to report these, though?

There are two scenarios here. The first is mixed TS/JS, where people have legacy JS as part of the compilation via allowJs. These people are running `tsc` regularly, and I don't want to block their build on a bogus error.

@DanielRosenwasser 
The other scenario is pure JS users who never run tsc. I believe it's these people that @DanielRosenwasser is thinking with
> I do think treading lightly might be the way to go

I agree. Shipping this as a suggestion at first is a good start. Based on the time I added all implicit any errors as suggestions, I think nobody on VS or Code will notice. Then styling them differently is a question of whether TS provides the errors in a separate function or whether the editor filters the suggestion list. I think either is fine but leaving it to the editors is certainly easier for me. =)

**Comment by sandersn** [comment]
Hi all, this is ready to review again. It now logs all errors as suggestions in unchecked JS files. The filtering in program.ts is gone entirely since suggestions are handled separately for normal errors.

**Review by uniqueiniquity — APPROVED** [review]
I'm much happier from the consumer side for these to just be suggestion diags.

**Comment by sandersn** [comment]
After talking to @DanielRosenwasser, I'm going to merge this as-is, with the error messages as slight variants of the originals. 

As I add errors to plain JS, better error display will become more important &mdash; for this PR, errors are hardly visible with Code's default suggestion UI &mdash; so we're going to investigate improved type display next. As part of that, I expect to learn how the error messages should change.
